### ISSUE TITLE
Selector - Check for True/False

### ISSUE THREAD

**Issue by TZwielehner** [IfcOpenShell/IfcOpenShell#2082]
How can I select an element with a boolean property? 

`.IfcWall[Pset_WallCommon.LoadBearing = "False"]`

I think I tried every possible combination... As far as I understand the syntax all checks are done on a "string" basis, right?

**Comment by aothms** [comment]
It certainly looks like that:

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/selector.py#L38

filter_value is set directly to escaped_string

When comparing, left hand side or right hand side is coerced based on the operand type:

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/selector.py#L210

filter_value instead could be something like:

SIGNED_NUMBER|ESCAPED_STRING|BOOLEAN

With BOOLEAN "True"|"False"|"Unknown"

And then adjusting filter_element to not blindly cast but take into account the value type from the grammar.

**Comment by Moult** [comment]
In addition to fixing this, I'd like to take a fresh look at the selector syntax. I think it should be revised to:

 1. Allow chaining of selectors, right now it always parses the entire dataset
 2. Follow the schema by IfcReference - that way this isn't just a bespoke IfcOpenShell query language, it actually follows the IFC spec

Would you be interested in hacking on this?

**Comment by Moult** [comment]
I did a little experiment to see what would happen if the IFC query syntax followed IfcReference. Let's recreate this:

```
.IfcWall[Pset_WallCommon.LoadBearing = "False"]
```

Here I've taking inspiration from IfcReference where an indentation means an inner reference, and an inner reference simply means "filter parent element" instead of "select subelement". IfcReference only allows pointing to an element and its attribute, but not giving a criteria to filter by, so let's also imagine IfcReference was enhanced with filtration criteria (=, <, >, in, etc)

```
 - TypeIdentifier: IfcWall
 - AttributeIdentifier: IsDefinedBy
     - TypeIdentifier: IfcRelDefinesByProperties
     - AttributeIdentifier: RelatingPropertyDefinition
         - TypeIdentifier: IfcPropertySet
         - AttributeIdentifier: Name="Pset_WallCommon"
         - AND
         - TypeIdentifier: IfcPropertySet
         - AttributeIdentifier: HasProperties
             - AttributeIdentifier: Name="LoadBearing"
             - AND
             - AttributeIdentifier: NominalValue=False
```

Naturally it's ridiculously verbose but that's just how IfcReference works - it literally lets you pinpoint exactly the relationship you're after.

Here's some randomly invented shorthand syntax:

```
.IfcWall[IsDefinedBy(.IfcRelDefinesByProperties[RelatingPropertyDefinition([Name="Pset_WallCommon"&HasProperties([Name="LoadBearing"&NominalValue=False]))])])]
```

Cool. So in theory you can invent any filter you want.

Except that it looks horrific and nobody would ever want to type that.

In conclusion I was wrong and it is a bad idea to turn IfcReference into a syntax. Even the simplest of relationships becomes complex so even if we provide shorthands for common relationships (e.g. property filters) I don't think it adds any value.

**Comment by Moult** [comment]
So this is now fixed, but I'll probably spend a bit more time doing some long-overdue upgrades on the selector to support things like chaining selections and "in" or "not" comparisons.

**Comment by aothms** [comment]
Great work. I'm slightly less appalled by the horrific query. Graph/tree queries are just ugly per definition. 

You could do the same trick as the semantic building data crowd did. In principal the syntax is the horrific query, but you allow certain shortcuts (graph rewriting on the query ast). It's overkill though. Maybe we should come up with an "academic issue list" or something. "Fun" issues that are nicely scoped to a MSc thesis that we distribute to our academic friends.

https://github.com/IfcOpenShell/IfcOpenShell/labels/academic

**Comment by Moult** [comment]
Cheers, just a heads up that in the past series of commits the selector has had a number of upgrades.

**Comment by pierremonico** [comment]
I am absolutely delighted to find this issue. I started digging into the `utils.selector` class a few weeks back and wanted to suggest a few things, among which boolean support. Let me know if I am polluting this issue but here are my two cents:
- I find the query capabilities extremely powerful and have at least two (real) projects on which I could make use of it immediately
- A few things are a bit off, i.e. you can't include `-` in property names. I quickly patched it myself but I am a total noob in Lark so not sure how sustainable it is 
- A few other things would be very helpful, in particular checking for simple existence of things like a property (**not** its value)
- I am blown away by your comment about chaining @Moult. I needed this to process intersection between elements matching queries, but have to rely on joining the queries with `&` and rerunning them on the all elements for now

Again, sorry if I polluted this issue but I thought it would be helpful to chime in with some of the use-cases we have and some of the features they could make use of.

I can:
- Give you more insights on what exactly we are doing/how we use `selector`
- Help hack on it but I'd probably need a little guidance at least on 1. what Lark docs to refer to and 2. what the general implementation should look like

**Comment by aothms** [comment]
> I quickly patched it myself but I am a total noob in Lark so not sure how sustainable it is

I'd say, send in a Pull Request and we discuss from there.

> checking for simple existence of things 

~~~
filter: "[" filter_key (comparison filter_value)? "]"
~~~

We currently have this ^ as the filter expression. So it already allows `.IfcWall[IsExternal]` I'm not sure what it means. Maybe @Moult  can clarify.

I'd say it's not unthinkable if we broaden the expression to other things than binary comparison. So filter becomes `"[" expression "]"` and `expression: binary_op | unary_op`, `binary_op: filter_key (comparison filter_value)?`, `unary_op: "exists" filter_key`. For example. Took a few shortcuts. We'd have to think about how negation of `exists x`  works, etc.

**Comment by Moult** [comment]
More than happy to get a PR on this. I thought I already added support for checking for existence, but perhaps not, so the latest commit now adds that support. Typing this after I reread @aothms' comment maybe I rushed it and the "not exists" doesn't quite work hmmm.

Chaining works like this (maybe not syntactically so beautiful but it works):

```python
results = ifcopenshell.util.selector.Selector(ifc, '.IfcWall')
subresults = ifcopenshell.util.selector.Selector(ifc, '.IfcWall', elements=results)
```

**Comment by aothms** [comment]
> Chaining works like this 

You can also use python built in expressions on set:

~~~py
set(ifcopenshell.util.selector.Selector().parse(f, '.IfcWall')) | set(ifcopenshell.util.selector.Selector().parse(f, '.IfcSlab'))
~~~

**Comment by pierremonico** [comment]
Thanks for the prompt replies. 

- For the `-` support I can send a PR but essentially it's a one character edit. 
- I will dig into the existence checking capabilities 
- Thanks for the hint on chaining, I'll also try that out 

All in all I think that I will run a few more tests on real-world use-cases and take some notes + write documentation. I will then post both here and maybe the docs can land in the OSArch wiki. Again, I truly believe that this sort of querying is very powerful.

**Comment by Moult** [comment]
I just realised potentially .IfcWall[Foo] is equivalent to .IfcWall[Foo!=NULL], so the opposite of .IfcWall[Foo] is .IfcWall[Foo=NULL]
